php/libraries/PipelineArchive.class.inc

```diff
@@ -0,0 +1,83 @@
+<?php declare(strict_types=1);
+
+use \Psr\Http\Message\ResponseInterface;
+use \Psr\Http\Message\ServerRequestInterface;
+use \LORIS\Http\Endpoint;
+
+/**
+ * This class handles pipeline archives downlaod. It should
+ * serve as the ONLY way that files are downloaded so that all file downloading
+ * functionality can occur on a well-tested foundation.
+ *
+ * @license http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ */
+class PipelineArchive extends Endpoint
+{
+    const ERROR_EMPTY_FILENAME = 'Invalid filename: cannot be empty';
+    const ERROR_FILE_NOT_FOUND = 'File requested cannot be found';
+
+    /**
+     * Given an HTTP request, serve the file to the client.
+     *
+     * @param ServerRequestInterface $request An HTTP Request that contains files.
+     *
+     * @return ResponseInterface
+     */
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        if ($request->getMethod() != "GET") {
+            return new \LORIS\Http\Response\JSON\MethodNotAllowed(
+                ['GET']
+            );
+        }
+
+        $filename = strval($request->getQueryParams()["filename"]);
+        if (empty($filename)) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                self::ERROR_EMPTY_FILENAME
+            );
+        }
+
+        $pipeline = strval($request->getQueryParams()["pipeline"]);
+        $version = strval($request->getQueryParams()["version"]);
+
+        $path_parts = pathinfo($filename);
+        $ext = $path_parts['extension'];
+        switch ($ext) {
+            case 'zip':
+                $mime = "application/x-zip";
```

Comment:
I don't think this mime type is correct. zip (without x-) is a standardized mime type.